python_modules/dagster/dagster/scheduler/scheduler.py

```diff
@@ -106,11 +108,83 @@ def launch_scheduled_runs(
     check.inst_param(instance, "instance", DagsterInstance)
     check.inst_param(workspace, "workspace", IWorkspace)
 
-    schedules = [
-        s
-        for s in instance.all_stored_job_state(job_type=InstigatorType.SCHEDULE)
-        if s.status == InstigatorStatus.RUNNING
-    ]
+    workspace_snapshot = {
+        location_entry.origin: location_entry
+        for location_entry in workspace.get_workspace_snapshot().values()
+    }
+
+    all_schedule_states = {
+        schedule_state.origin.get_id(): schedule_state
+        for schedule_state in instance.all_stored_job_state(job_type=InstigatorType.SCHEDULE)
+    }
+
+    schedules = {}
+    for location_entry in workspace_snapshot.values():
+        repo_location = location_entry.repository_location
+        if repo_location:
+            for repo in repo_location.get_repositories().values():
+                for schedule in repo.get_external_schedules():
+                    origin_id = schedule.get_external_origin().get_id()
+                    if schedule.get_current_instigator_state(
+                        all_schedule_states.get(origin_id)
+                    ).is_running:
+                        schedules[origin_id] = schedule
+        elif location_entry.load_error and log_verbose_checks:
+            logger.warning(
+                f"Could not load location {location_entry.origin.location_name} to check for schedules due to the following error: {location_entry.load_error}"
+            )
+
+    # Remove any schedule states that were previously created with AUTOMATICALLY_RUNNING
+    # and can no longer be found in the workspace (so that if they are later added
+    # back again, their timestamps will start at the correct place)
```

Comment:
zesty - do you feel confident in this or does it merit some discussion?

Reply:
open to alternatives, it's tested under test_change_default_status